state/redis.c

```diff
@@ -150,45 +178,107 @@ void object_table_get_entry(redisAsyncContext *c, void *r, void *privdata) {
-  }
   node_id node = *task_instance_node(task_instance);
   int32_t state = *task_instance_state(task_instance);
-  redisAsyncCommand(db->context, NULL, NULL, "PUBLISH task_log:%b:%d %b",
-                    (char *) &node.id[0], UNIQUE_ID_SIZE, state,
-                    (char *) task_instance, task_instance_size(task_instance));
-  if (db->context->err) {
-    LOG_REDIS_ERR(db->context, "error publishing task in task_log_add_task");
+  
+  LOG_DEBUG("Called log_publish callback");
+
+  /* Check whether the vector (requests_info) indicating the status of the requests has been allocated.
+   * If was not allocate it, allocate it and initialize it.
+   * This vector has an entry for each redis command, and it stores true if a reply for that command
+   * has been received, and false otherwise.
+   * The first entry in the callback corresponds to RPUSH, and the second entry to PUBLISH.
+   */
+#define NUM_DB_REQUESTS 2
+#define PUSH_INDEX      0
+#define PUBLISH_INDEX   1
+  if (cb_data->requests_info == NULL) {
+    cb_data->requests_info = malloc(NUM_DB_REQUESTS * sizeof(bool));
+    for (int i = 0; i < NUM_DB_REQUESTS; i++) {
+      ((bool *)cb_data->requests_info)[i] = false;
+    }
+  }
+
+  if (((bool *)cb_data->requests_info)[PUSH_INDEX] == false) {
+    if (*task_instance_state(task_instance) == TASK_WAITING) {
+      redisAsyncCommand(db->context, redis_task_log_publish_push_cb, cb_data, "RPUSH tasklog:%b %b",
+                        (char *) &task_iid.id[0], UNIQUE_ID_SIZE,
+                        (char *) task_instance, task_instance_size(task_instance));
+    } else {
+      task_update update = {.state = state, .node = node};
+      redisAsyncCommand(db->context, redis_task_log_publish_push_cb, cb_data, "RPUSH tasklog:%b %b",
+                        (char *) &task_iid.id[0], UNIQUE_ID_SIZE,
+                        (char *) &update, sizeof(update));
+    }
+
+    if (db->context->err) {
+      LOG_REDIS_ERR(db->context, "error setting task in task_log_add_task");
+    }
+  }
+
+  if (((bool *)cb_data->requests_info)[PUBLISH_INDEX] == false) {
+    redisAsyncCommand(db->context, redis_task_log_publish_publish_cb, cb_data, "PUBLISH task_log:%b:%d %b",
+                      (char *) &node.id[0], UNIQUE_ID_SIZE, state,
+                      (char *) task_instance, task_instance_size(task_instance));
+
+    if (db->context->err) {
+      LOG_REDIS_ERR(db->context, "error publishing task in task_log_add_task");
+    }
+  }
+}
+
+void redis_task_log_publish_push_cb(redisAsyncContext *c, void *r, void *privdata) {
+  db_handle *db = c->data;
+  table_callback_data *cb_data = privdata;
+  redisReply *reply = r;
+
+  CHECK(cb_data->requests_info != NULL);
+  ((bool *)cb_data->requests_info)[PUSH_INDEX] = true;
+
+  if (((bool *)cb_data->requests_info)[PUBLISH_INDEX] == true) {
+    if (cb_data->done_cb) {
+      task_log_done_cb done_cb = cb_data->done_cb;
+      done_cb(cb_data->id, cb_data->user_context);
```

Comment:
We should remove the timer here to prevent the failure callback from being called (see comment in state/table.c)